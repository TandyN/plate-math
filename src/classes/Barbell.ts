import Weight from './Weight'

class Barbell {
  barbell_weight: number
  weights: Array<Weight>

  constructor(barbell_weight = 20) {
    this.barbell_weight = barbell_weight
    this.weights = []
  }
}
