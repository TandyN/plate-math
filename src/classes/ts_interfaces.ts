enum WeightType {
  kgs = 'kilograms',
  lbs = 'pounds',
}

export { WeightType }
