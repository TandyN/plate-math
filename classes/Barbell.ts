import { Plate } from './Plate'
import { WeightType } from './ts_interfaces'

class Barbell {
  #multiplier: number
  #weight: number // will always be in kilograms
  #attached_plates: Array<Plate>

  constructor(weight = 20, weight_type: WeightType = WeightType.kgs) {
    this.#multiplier = 2.205 // kg --> lbs
    this.#weight = weight
    this.set_weight(weight, weight_type)
    this.#attached_plates = []
  }

  get_weight(): number {
    return this.#weight
  }

  set_weight(weight: number, weight_type: WeightType = WeightType.kgs): void {
    if (weight < 0) {
      throw new Error('barbell weight can not be negative')
    }

    let input_weight = weight

    if (weight_type === WeightType.lbs) {
      input_weight /= this.#multiplier
    } else if (weight_type !== WeightType.kgs) {
      console.warn(
        `weight_type '${weight_type}' does not exist for setting barbell weight. assuming kilograms`,
      )
    }

    this.#weight = input_weight
  }

  get_attached_plates(): Array<Plate> {
    return this.#attached_plates
  }

  add_plate(weight: number, weight_type: WeightType = WeightType.kgs): void {
    // if weight type doesn't exist
    if (!Object.values(WeightType).includes(weight_type)) {
      console.warn(
        `weight_type '${weight_type}' does not exist for adding plate. assuming kilograms`,
      )
      weight_type = WeightType.kgs
    }
    const added_plate: Plate = new Plate(weight, weight_type)
    this.#attached_plates.push(added_plate)
  }

  remove_last_plate() {
    const removed_plate: Plate | undefined = this.#attached_plates.pop()
    return removed_plate
  }

  remove_all_plates() {
    const current_attached_plates: Array<Plate> = this.#attached_plates
    this.#attached_plates = []
    return current_attached_plates
  }

  get_total_weight(weight_type: WeightType = WeightType.kgs): number {
    const bar_weight =
      weight_type === WeightType.lbs
        ? this.#weight * this.#multiplier
        : this.#weight

    return (
      this.#attached_plates.reduce((current_total, weight_node): number => {
        let current_weight: number = weight_node.weight
        if (
          weight_type === WeightType.lbs &&
          weight_node.weight_type !== weight_type
        ) {
          current_weight *= this.#multiplier
        } else if (
          weight_type === WeightType.kgs &&
          weight_node.weight_type !== weight_type
        ) {
          current_weight /= this.#multiplier
        }
        return current_total + current_weight
      }, 0) *
        2 +
      bar_weight
    )
  }

  fill_to_target(
    target_weight: number, // What they want the weight to be
    target_weight_type: WeightType = WeightType.kgs, // The weight type they specify
    target_plate_type: WeightType = WeightType.kgs, // The plate type they want to put on
  ): Array<Plate> {
    let use_plate_type: Array<[number, number]>
    let current_weight: number

    if (target_plate_type === WeightType.lbs) {
      use_plate_type = [
        [45, 45], // [Used for Plate Object, Used for calculation]
        [35, 35],
        [25, 25],
        [10, 10],
        [5, 5],
        [2.5, 2.5],
        [1.25, 1.25],
      ]
    } else {
      use_plate_type = [
        [25, 25],
        [20, 20],
        [15, 15],
        [10, 10],
        [5, 5],
        [2.5, 2.5],
        [1.25, 1.25],
        [0.5, 0.5],
        [0.25, 0.25],
      ]
    }

    if (target_weight_type === WeightType.lbs) {
      current_weight = this.get_total_weight(WeightType.lbs)
      if (target_weight_type !== target_plate_type) {
        use_plate_type = use_plate_type.map((plate_weight) => [
          plate_weight[0],
          plate_weight[0] * this.#multiplier,
        ])
      }
    } else {
      current_weight = this.get_total_weight(WeightType.kgs)
      if (target_weight_type !== target_plate_type) {
        use_plate_type = use_plate_type.map((plate_weight) => [
          plate_weight[0],
          plate_weight[0] / this.#multiplier,
        ])
      }
    }

    const target_attached_weights: Array<Plate> = []
    let plate_index = 0

    while (plate_index < use_plate_type.length) {
      const added_weight = current_weight + use_plate_type[plate_index][1] * 2

      if (added_weight <= target_weight) {
        current_weight = added_weight
        target_attached_weights.push(
          new Plate(use_plate_type[plate_index][0], target_plate_type),
        )
        this.#attached_plates.push(
          new Plate(use_plate_type[plate_index][0], target_plate_type),
        )
      } else {
        plate_index++
      }
    }

    return target_attached_weights
  }
}

export { Barbell }
