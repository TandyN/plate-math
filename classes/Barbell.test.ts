import { Barbell } from './Barbell'
import { WeightType } from './ts_interfaces'

// kg to lbs multiplier
const multiplier = 2.205

afterEach(() => {
  jest.clearAllMocks()
})

describe('Barbell Class', () => {
  describe('Getting/Setting Barbell Weight', () => {
    it('should output weight to kilograms if kilograms is specified', () => {
      let barbell = new Barbell(45, WeightType.kgs)
      expect(barbell.get_weight()).toBe(45)

      barbell = new Barbell(0, WeightType.kgs)
      expect(barbell.get_weight()).toBe(0)

      barbell = new Barbell(59.2, WeightType.kgs)
      expect(barbell.get_weight()).toBe(59.2)
    })

    it('should convert barbell weight to kilograms if weight type pounds is specified', () => {
      let barbell = new Barbell(45, WeightType.lbs)
      expect(barbell.get_weight()).toBe(45 / multiplier)

      barbell = new Barbell(0, WeightType.lbs)
      expect(barbell.get_weight()).toBe(0)

      barbell = new Barbell(59.2, WeightType.lbs)
      expect(barbell.get_weight()).toBe(59.2 / multiplier)
    })

    it('should log a warning and assume barbell weight type is kilograms if weight type does not exist', () => {
      const console_warn_spy = jest.spyOn(console, 'warn')

      let barbell = new Barbell(33, 'fake weight type' as never)
      expect(barbell.get_weight()).toBe(33)

      barbell = new Barbell(0, 'fake weight type' as never)
      expect(barbell.get_weight()).toBe(0)

      barbell = new Barbell(59.2, 'fake weight type' as never)
      expect(barbell.get_weight()).toBe(59.2)

      expect(console_warn_spy.mock.calls).toMatchSnapshot()
    })

    it('should default barbell weight to 20 kilograms if no arguments', () => {
      const barbell = new Barbell()
      expect(barbell.get_weight()).toBe(20)
    })

    it('should set barbell weight to specified weight', () => {
      const barbell = new Barbell()
      barbell.set_weight(25, WeightType.kgs)
      expect(barbell.get_weight()).toBe(25)

      barbell.set_weight(45, WeightType.lbs)
      expect(barbell.get_weight()).toBe(45 / multiplier)
    })

    it('should set assume weight type of kilograms if no weight type specified', () => {
      const barbell = new Barbell(21)
      expect(barbell.get_weight()).toBe(21)
      barbell.set_weight(31, WeightType.kgs)
      expect(barbell.get_weight()).toBe(31)
    })

    it('should log a warning and set assume weight type of kilograms if specified weight type doesn not exist', () => {
      const console_warn_spy = jest.spyOn(console, 'warn')

      const barbell = new Barbell(21)
      expect(barbell.get_weight()).toBe(21)

      barbell.set_weight(31, 'fake weight type' as never)
      expect(barbell.get_weight()).toBe(31)

      expect(console_warn_spy.mock.calls).toMatchSnapshot()
    })

    it('should throw an error if user tries to make the barbell weight negative by initalizing or setting', () => {
      expect(() => {
        new Barbell(-20)
      }).toThrowErrorMatchingSnapshot()

      expect(() => {
        new Barbell(-20, WeightType.lbs)
      }).toThrowErrorMatchingSnapshot()

      const barbell = new Barbell()

      expect(() => {
        barbell.set_weight(-50, WeightType.kgs)
      }).toThrowErrorMatchingSnapshot()
    })
  })

  describe('Getting/Setting Attached Plates', () => {
    describe('add_plate function', () => {
      it('should add to attached_plates if weight number exists as a kilogram plate', () => {
        const barbell = new Barbell()
        barbell.add_plate(10, WeightType.kgs)

        expect(barbell.get_attached_plates().length).toBe(1)
        expect(barbell.get_attached_plates()[0].get_weight()).toBe(10)
        expect(barbell.get_attached_plates()[0].get_weight_type()).toBe(
          'kilograms',
        )

        barbell.add_plate(15, WeightType.kgs)
        expect(barbell.get_attached_plates().length).toBe(2)
        expect(barbell.get_attached_plates()[1].get_weight()).toBe(15)
        expect(barbell.get_attached_plates()[1].get_weight_type()).toBe(
          'kilograms',
        )
      })

      it('should add to attached_plates if weight number exists as a pound plate', () => {
        const barbell = new Barbell()

        barbell.add_plate(10, WeightType.lbs)
        expect(barbell.get_attached_plates().length).toBe(1)
        expect(barbell.get_attached_plates()[0].get_weight()).toBe(10)
        expect(barbell.get_attached_plates()[0].get_weight_type()).toBe(
          'pounds',
        )

        barbell.add_plate(25, WeightType.lbs)
        expect(barbell.get_attached_plates().length).toBe(2)
        expect(barbell.get_attached_plates()[1].get_weight()).toBe(25)
        expect(barbell.get_attached_plates()[1].get_weight_type()).toBe(
          'pounds',
        )
      })
    })

    describe('remove_last_plate function', () => {
      it('should remove plate from attached_plates', () => {
        const barbell = new Barbell()

        barbell.add_plate(10, WeightType.lbs)
        barbell.add_plate(5, WeightType.lbs)
        expect(barbell.get_attached_plates().length).toBe(2)

        const first_removed_plate = barbell.remove_last_plate()
        expect(first_removed_plate?.get_weight()).toBe(5)
        expect(barbell.get_attached_plates().length).toBe(1)

        const second_removed_plate = barbell.remove_last_plate()
        expect(second_removed_plate?.get_weight()).toBe(10)
        expect(barbell.get_attached_plates().length).toBe(0)
      })

      it('should return undefined if there are no plates to remove', () => {
        const barbell = new Barbell()
        expect(barbell.get_attached_plates().length).toBe(0)

        const removed_plate = barbell.remove_last_plate()
        expect(removed_plate).toBeUndefined()
        expect(barbell.get_attached_plates().length).toBe(0)
      })
    })

    describe('remove_all_plates function', () => {
      it('should remove all plates from attached_plates', () => {
        const barbell = new Barbell()

        barbell.add_plate(10, WeightType.lbs)
        barbell.add_plate(10, WeightType.lbs)
        expect(barbell.get_attached_plates().length).toBe(2)

        let removed_plates = barbell.remove_all_plates()
        expect(removed_plates.length).toBe(2)
        expect(barbell.get_attached_plates().length).toBe(0)

        removed_plates = barbell.remove_all_plates()
        expect(removed_plates.length).toBe(0)
        expect(barbell.get_attached_plates().length).toBe(0)
      })
    })
  })

  describe('get_total_weight function', () => {
    it("should calculate all plates attached to barbell in pounds if 'pounds' is passed as argument", () => {
      const barbell = new Barbell()

      // just the bar
      expect(barbell.get_total_weight(WeightType.lbs)).toBe(20 * multiplier)

      barbell.add_plate(10, WeightType.kgs)
      barbell.add_plate(10, WeightType.lbs)
      barbell.add_plate(10, WeightType.kgs)
      barbell.add_plate(10, WeightType.lbs)
      barbell.add_plate(5, WeightType.kgs)

      const actual =
        Math.round(barbell.get_total_weight(WeightType.lbs) * 100) / 100

      // add kg plates, multiple by 2 for both sides of the barbell, convert to lbs, add barbell weight, then add pound lb at the end for both sides
      const expected = 25 * 2 * multiplier + 20 * multiplier + 40

      expect(actual).toBe(Math.round(expected * 100) / 100)
    })

    it("should calculate all plates attached to barbell in kilograms if 'kilograms' is passed as argument", () => {
      const barbell = new Barbell()

      // just the bar
      expect(barbell.get_total_weight(WeightType.kgs)).toBe(20)

      barbell.add_plate(10, WeightType.lbs)
      barbell.add_plate(10, WeightType.kgs)
      barbell.add_plate(10, WeightType.lbs)
      barbell.add_plate(10, WeightType.kgs)
      barbell.add_plate(5, WeightType.lbs)

      const actual =
        Math.round(barbell.get_total_weight(WeightType.kgs) * 100) / 100

      // add lb plates, multiple by 2 for both sides of the barbell, convert to kgs, add barbell weight, then add kg plates at the end for both sides
      const expected = (25 * 2) / multiplier + 20 + 40

      expect(actual).toBe(Math.round(expected * 100) / 100)
    })

    it("should assume weight type is kgs if 'pounds' is not the explicit weight type", () => {
      // bar defaults to 20kg
      const barbell = new Barbell()

      expect(barbell.get_total_weight(WeightType.kgs)).toBe(20)
      expect(barbell.get_total_weight('test' as never)).toBe(20)
    })
  })

  describe('fill_to_target function', () => {
    it('should fill the attached_plates as close to specified weight as possible with lb plates', () => {
      const barbell = new Barbell(45, WeightType.lbs)

      barbell.add_plate(45, WeightType.lbs)

      const attached_plates = barbell.fill_to_target(
        318,
        WeightType.lbs,
        WeightType.lbs,
      )

      expect(attached_plates.length).toBe(3)
      expect(attached_plates[0].get_weight()).toBe(45)
      expect(attached_plates[1].get_weight()).toBe(45)
      expect(attached_plates[2].get_weight()).toBe(1.25)

      expect(barbell.get_attached_plates().length).toBe(4)
      expect(barbell.get_attached_plates()[1].get_weight()).toBe(45)
      expect(barbell.get_attached_plates()[2].get_weight()).toBe(45)
      expect(barbell.get_attached_plates()[3].get_weight()).toBe(1.25)

      expect(barbell.get_total_weight(WeightType.lbs)).toBe(317.5)
    })

    it('should fill the attached_plates as close to specified weight as possible with kg plates', () => {
      const barbell = new Barbell()

      barbell.add_plate(25, WeightType.kgs)

      const attached_plates = barbell.fill_to_target(
        170.6,
        WeightType.kgs,
        WeightType.kgs,
      )

      expect(attached_plates.length).toBe(3)
      expect(attached_plates[0].get_weight()).toBe(25)
      expect(attached_plates[1].get_weight()).toBe(25)
      expect(attached_plates[2].get_weight()).toBe(0.25)

      expect(barbell.get_attached_plates().length).toBe(4)
      expect(barbell.get_attached_plates()[1].get_weight()).toBe(25)
      expect(barbell.get_attached_plates()[2].get_weight()).toBe(25)
      expect(barbell.get_attached_plates()[3].get_weight()).toBe(0.25)

      expect(barbell.get_total_weight(WeightType.kgs)).toBe(170.5)
    })

    it('should fill the attached_plates as close to specified weight as possible with kg plates given the target weight in lbs', () => {
      const barbell = new Barbell()

      barbell.add_plate(25, WeightType.kgs)

      const attached_plates = barbell.fill_to_target(
        376.1,
        WeightType.lbs,
        WeightType.kgs,
      )

      expect(attached_plates.length).toBe(3)
      expect(attached_plates[0].get_weight()).toBe(25)
      expect(attached_plates[1].get_weight()).toBe(25)
      expect(attached_plates[2].get_weight()).toBe(0.25)

      expect(barbell.get_attached_plates().length).toBe(4)
      expect(barbell.get_attached_plates()[1].get_weight()).toBe(25)
      expect(barbell.get_attached_plates()[2].get_weight()).toBe(25)
      expect(barbell.get_attached_plates()[3].get_weight()).toBe(0.25)

      expect(barbell.get_total_weight(WeightType.kgs)).toBe(170.5)
    })

    it('should fill the attached_plates as close to specified weight as possible with lb plates given the target weight in kgs', () => {
      const barbell = new Barbell(45, WeightType.lbs)

      barbell.add_plate(45, WeightType.lbs)

      const attached_plates = barbell.fill_to_target(
        144.2,
        WeightType.kgs,
        WeightType.lbs,
      )

      expect(attached_plates.length).toBe(3)
      expect(attached_plates[0].get_weight()).toBe(45)
      expect(attached_plates[1].get_weight()).toBe(45)
      expect(attached_plates[2].get_weight()).toBe(1.25)

      expect(barbell.get_attached_plates().length).toBe(4)
      expect(barbell.get_attached_plates()[1].get_weight()).toBe(45)
      expect(barbell.get_attached_plates()[2].get_weight()).toBe(45)
      expect(barbell.get_attached_plates()[3].get_weight()).toBe(1.25)

      expect(barbell.get_total_weight(WeightType.lbs)).toBe(317.5)
    })
  })
})
