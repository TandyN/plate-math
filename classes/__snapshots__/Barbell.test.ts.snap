// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Barbell Class Getting/Setting Attached Plates add_plate function should log a warning and assume weight type of kilograms if weight type does not exist 1`] = `
Array [
  Array [
    "weight_type 'fake weight type' does not exist for adding plate. assuming kilograms",
  ],
]
`;

exports[`Barbell Class Getting/Setting Attached Plates add_plate function should throw an error if number does not exist for kilogram plates 1`] = `"plate number '11' does not exist for 'kilograms'"`;

exports[`Barbell Class Getting/Setting Attached Plates add_plate function should throw an error if number does not exist for pound plates 1`] = `"plate number '11' does not exist for 'pounds'"`;

exports[`Barbell Class Getting/Setting Barbell Weight should log a warning and assume barbell weight type is kilograms if weight type does not exist 1`] = `
Array [
  Array [
    "weight_type 'fake weight type' does not exist for setting barbell weight. assuming kilograms",
  ],
  Array [
    "weight_type 'fake weight type' does not exist for setting barbell weight. assuming kilograms",
  ],
  Array [
    "weight_type 'fake weight type' does not exist for setting barbell weight. assuming kilograms",
  ],
]
`;

exports[`Barbell Class Getting/Setting Barbell Weight should log a warning and set assume weight type of kilograms if specified weight type doesn not exist 1`] = `
Array [
  Array [
    "weight_type 'fake weight type' does not exist for setting barbell weight. assuming kilograms",
  ],
]
`;

exports[`Barbell Class Getting/Setting Barbell Weight should throw an error if user tries to make the barbell weight negative by initalizing or setting 1`] = `"barbell weight can not be negative"`;

exports[`Barbell Class Getting/Setting Barbell Weight should throw an error if user tries to make the barbell weight negative by initalizing or setting 2`] = `"barbell weight can not be negative"`;

exports[`Barbell Class Getting/Setting Barbell Weight should throw an error if user tries to make the barbell weight negative by initalizing or setting 3`] = `"barbell weight can not be negative"`;
